src/darkleaf/effect/core.cljc

```diff
@@ -19,37 +19,18 @@
      (cr {! i/coeffect} ~@body)
      :coroutine))
 
-(defn- update-head [coll f & args]
-  (if (seq coll)
-    (-> coll
-        (pop)
-        (conj (apply f (peek coll) args)))
-    coll))
+(defn call [coroutine return coeffect]
+  (let [val (i/with-coeffect coeffect coroutine)]
+    (case (i/kind val)
+      :effect [val (partial coroutine call return)]
+      :coroutine (val call (partial coroutine call return) nil)
+      :wrapped (recur coroutine return (first val))
+      (return val))))
 
-(defn- stack->continuation [stack]
-  (fn [coeffect]
-    (loop [stack    stack
-           coeffect coeffect]
-      (if (empty? stack)
-        [coeffect nil]
-        (let [stack     (update-head stack (fn clone [mutable-coroutine]
-                                             (mutable-coroutine identity)))
-              coroutine (peek stack)
-              val       (i/with-coeffect coeffect coroutine)]
-          (case (i/kind val)
-            :effect    [val (stack->continuation stack)]
-            :coroutine (recur (conj stack val) nil)
-            :wrapped   (recur  stack (first val))
-            ;; coroutine is finished
-            (recur (pop stack) val)))))))
+(defn done [x] [x nil])
 
 (defn continuation [effn]
-  (fn [args]
-    (let [coroutine (apply effn args)
-          stack     (list coroutine)
-          cont      (stack->continuation stack)
-          coeffect  ::not-used]
-      (cont coeffect))))
+  (fn [args] ((apply effn args) call done nil)))
```

Comment:
There are two functions:
1. effn - a function that returns coroutine
2. continuation - a function that builds continuation from effn

Cloning coroutine by `((apply effn args) call done nil)` doesn't make sense because continuation always gets a fresh coroutine.

I clone stack because anyone may call a continuation several times. I think about middleware that provide block/return-from semantics, so effect library should provide continuations without internal mutable state.

Reply:
I will be happy if cloroutine provides an immutable api. I.e:
```clojure
(let [coroutine ...
        [value coroutine] (coroutine)])
```

Reply:
> Cloning coroutine by ((apply effn args) call done nil) doesn't make sense because continuation always gets a fresh coroutine.

Yes. If you can guarantee `eff` blocks are not reused, the coroutine is always fresh so you can avoid cloning and do `(call (apply effn args) done nil)` instead.

> I clone stack because anyone may call a continuation several times. I think about middleware that provide block/return-from semantics, so effect library should provide continuations without internal mutable state.

I totally understand why you would want to expose an immutable API. That's why cloning is required in the first place when calling the resuming function. What I was pointing out is, in the special case of `:wrapped`, we don't expose the current continuation so we can reuse the coroutine and avoid cloning.

> I will be happy if cloroutine provides an immutable api.

I chose a mutable api for cloroutine because I did not want to compromise on performance. If you need single-shot continuations only, you can avoid cloning and minimize allocations, and if you need multi-shot continuations you can work on a copy (basically what you did).

Reply:
> What I was pointing out is, in the special case of :wrapped, we don't expose the current continuation so we can reuse the coroutine and avoid cloning.

It's interesting but not obvious. I'll think about it.

> I chose a mutable api for cloroutine because I did not want to compromise on performance.

Understand.